doc/manpages/ccngroup.1
'\" t
.\"     Title: ccngroup
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNGROUP" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccngroup \- show and modify access control groups for access control of a CCNx content namespace\&.
.SH "SYNOPSIS"
.sp
\fBccngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-list
.sp
\fBccngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-listmembers \fIgroupFriendlyName\fR
.sp
\fBccngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-create \fIgroupFriendlyName\fR \fIgroupMember\fR\&...
.sp
\fBccngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-delete \fIgroupFriendlyName\fR
.sp
\fBccngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-add \fIgroupFriendlyName\fR \fIgroupMemberToAdd\fR\&...
.sp
\fBccngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-remove \fIgroupFriendlyName\fR \fIgroupMemberToRemove\fR\&...
.SH "DESCRIPTION"
.sp
The \fBccngroup\fR utility lists or modifies access control groups published with CCNx\&. Groups may be referenced in access control lists (ACLs) under the CCNx access control profile\&.
.sp
Groups have short, human\-readable "friendly" names by which they are identified with this tool, although the internal identifier for a group is a digest of a public key\&.
.SH "OPTIONS"
.PP
\fB\-as\fR \fIpathToKeystore\fR
.RS 4
Specifies a file path to a keystore; this is used for signing any newly\-created content\&. Applicable to all operations\&. The default keystore is used if this is not specified\&.
.RE
.PP
\fB\-list\fR
.RS 4
List the groups that exist\&.
.RE
.PP
\fB\-listmembers\fR \fIgroupFriendlyName\fR
.RS 4
List the members of the group
\fIgroupFriendlyName\fR\&.
.RE
.PP
\fB\-create\fR \fIgroupFriendlyName\fR \fIgroupMember\fR\&...
.RS 4
Create the group named
\fIgroupFriendlyName\fR
with initial members
\fIgroupMember\fR\&...
.RE
.PP
\fB\-delete\fR \fIgroupFriendlyName\fR
.RS 4
Delete the group named
\fIgroupFriendlyName\fR\&.
.RE
.PP
\fB\-add\fR \fIgroupFriendlyName\fR \fIgroupMemberToAdd\fR\&...
.RS 4
Add the
\fIgroupMemberToAdd\fR\&... members to the group
\fIgroupFriendlyName\fR\&.
.RE
.PP
\fB\-remove\fR \fIgroupFriendlyName\fR \fIgroupMemberToRemove\fR\&...
.RS 4
Remove the
\fIgroupMemberToRemove\fR\&... members from the group
\fIgroupFriendlyName\fR\&.
.RE
.PP
\fB\-debug\fR \fIportno\fR
.RS 4
Allows attachment of an eclipse remote debugger to the utility at the specified port number\&.
.RE
.PP
\fB\-javaopts\fR \fIoptions\fR
.RS 4
Allows use of arbitrary java properties or options with the utility\&.
.RE
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure (syntax or usage error; retrieval error)
.RE
.SH "AUTHOR"
.sp
Philippe Golle


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

javasrc/src/org/ccnx/ccn/utils/ccngroup.java
/*
 * A CCNx command line utility.
 *
 * Copyright (C) 2010, 2012 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

package org.ccnx.ccn.utils;

import java.io.File;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.SortedSet;
import java.util.logging.Level;

import org.ccnx.ccn.CCNHandle;
import org.ccnx.ccn.config.SystemConfiguration;
import org.ccnx.ccn.config.UserConfiguration;
import org.ccnx.ccn.impl.support.Log;
import org.ccnx.ccn.io.content.Link;
import org.ccnx.ccn.profiles.nameenum.EnumeratedNameList;
import org.ccnx.ccn.profiles.security.access.AccessDeniedException;
import org.ccnx.ccn.profiles.security.access.group.Group;
import org.ccnx.ccn.profiles.security.access.group.GroupAccessControlManager;
import org.ccnx.ccn.profiles.security.access.group.GroupManager;
import org.ccnx.ccn.profiles.security.access.group.MembershipListObject;
import org.ccnx.ccn.protocol.ContentName;

public class ccngroup {

	private static long TIMEOUT = 1000;
	private static ContentName userStorage = new ContentName(UserConfiguration.defaultNamespace(), "Users");
	private static ContentName groupStorage = new ContentName(UserConfiguration.defaultNamespace(), "Groups");

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// silence logging
		Log.setDefaultLevel(Level.WARNING);
		String extraUsage = "";

		if ((args == null) || (args.length == 0)) {
			usage(extraUsage);
		}

		int pos = 0;
		if (args[0].startsWith("[")) {
			extraUsage = args[0];
			pos++;
		}
		if (args[pos].equals("-h")) {
			usage(extraUsage);
		}
		if (args[pos].equals("-as")) {
			if (args.length < pos+2) usage(extraUsage);
			pos++;
			setUser(args[pos]);
			pos++;
		}

		if (args[pos].equals("-list")) {
			listGroups();
			System.exit(0);
		}
		else if (args[pos].equals("-listmembers")) {
			if (args.length < pos + 2) usage(extraUsage);
			pos++;
			String groupName = args[pos];
			listMembers(groupName);
			System.exit(0);
		}
		else if (args[pos].equals("-delete")) {
			if (args.length < pos + 2) usage(extraUsage);
			pos ++;
			String groupName = args[pos];
			deleteGroup(groupName);
			System.exit(0);
		}
		else if (args[pos].equals("-create") || args[pos].equals("-add") || args[pos].equals("-remove")) {
			if (args.length < pos + 2) usage(extraUsage);
			String command = args[pos];
			pos++;
			String groupName = args[pos];
			pos++;
			ArrayList<Link> groupMembers = new ArrayList<Link>();
			for (int i=pos; i<args.length; i++) {
				try {
					Link lk = new Link(ContentName.fromNative(args[i]));
					groupMembers.add(lk);
				}
				catch (Exception e) {
					e.printStackTrace();
					System.exit(1);
				}
			}
			if (command.equals("-create")) createGroup(groupName, groupMembers);
			else if (command.equals("-add")) addMember(groupName, groupMembers);
			else if (command.equals("-remove")) removeMember(groupName, groupMembers);
			System.exit(0);
		}
		else {
			usage(extraUsage);
		}

	}

	public static void usage(String extraUsage) {
		System.out.println("usage:");
		System.out.println("ccngroup " + extraUsage + "[-as pathToKeystore] -list");
		System.out.println("ccngroup " + extraUsage + "[-as pathToKeystore] -listmembers groupFriendlyName");
		System.out.println("ccngroup " + extraUsage + "[-as pathToKeystore] [-create | -add | -remove] groupFriendlyName (groupMember)*");
		System.out.println("ccngroup " + extraUsage + "[-as pathToKeystore] -delete groupFriendlyName");
		System.exit(1);
	}

	public static void setUser(String pathToKeystore) {
		File userDirectory = new File(pathToKeystore);
		String userConfigDir = userDirectory.getAbsolutePath();
		System.out.println("Loading keystore from: " + userConfigDir);
		UserConfiguration.setUserConfigurationDirectory(userConfigDir);
		// Assume here that the name of the file is the userName
		String userName = userDirectory.getName();
		if (userName != null) {
			System.out.println("User: " + userName);
			UserConfiguration.setUserName(userName);
		}
	}

	public static void listGroups() {
		try {
			EnumeratedNameList userDirectory = new EnumeratedNameList(groupStorage, CCNHandle.open());
			userDirectory.waitForChildren(TIMEOUT);
			Thread.sleep(TIMEOUT);

			SortedSet<ContentName> availableChildren = userDirectory.getChildren();
			if ((null == availableChildren) || (availableChildren.size() == 0)) {
				System.out.println("No group found in: " + groupStorage);
			}
			else {
				System.out.println(availableChildren.size() + " group(s) found in: " + groupStorage);
				for (ContentName child : availableChildren) {
					ContentName fullName = groupStorage.append(child);
					System.out.println(fullName);
				}
			}
		}
		catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void listMembers(String groupName) {
		try{
			GroupAccessControlManager acm = new GroupAccessControlManager(null, groupStorage, userStorage, CCNHandle.open());
			GroupManager gm = acm.groupManager();
			Thread.sleep(TIMEOUT);
			Group g = gm.getGroup(groupName, SystemConfiguration.getDefaultTimeout());
			MembershipListObject ml = g.membershipList();
			LinkedList<Link> lll = ml.contents();
			System.out.println("The group " + groupName + " has " + lll.size() + " members:");
			for (Link l: lll) {
				System.out.println(l.targetName());
			}
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
	}


	public static void createGroup(String groupName, ArrayList<Link> membersToAdd) {
		try {
			ContentName root = ContentName.fromNative("/");
			GroupAccessControlManager acm = new GroupAccessControlManager(root, groupStorage, userStorage, CCNHandle.open());
			GroupManager gm = acm.groupManager();
			gm.createGroup(groupName, membersToAdd, SystemConfiguration.getDefaultTimeout());
		} catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.out.println("Group " + groupName + " created, with members: ");
		for (Link lk: membersToAdd) {
			System.out.println(lk.targetName());
		}
	}

	public static void deleteGroup(String groupName) {
		try {
			GroupAccessControlManager acm = new GroupAccessControlManager(null, groupStorage, userStorage, CCNHandle.open());
			GroupManager gm = acm.groupManager();
			gm.deleteGroup(groupName);
		} catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.out.println("Deleted group " + groupName);
	}

	public static void addMember(String groupName, ArrayList<Link> membersToAdd) {
		try {
			GroupAccessControlManager acm = new GroupAccessControlManager(null, groupStorage, userStorage, CCNHandle.open());
			GroupManager gm = acm.groupManager();
			Thread.sleep(TIMEOUT);
			Group g = gm.getGroup(groupName, SystemConfiguration.getDefaultTimeout());
			g.modify(membersToAdd, null);
		}
		catch (AccessDeniedException aed) {
			System.out.println("You do not have the permission to edit the membership of Group " + groupName);
			System.exit(1);
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.out.println("Added to group " + groupName + " the following members: ");
		for (Link lk: membersToAdd) {
			System.out.println(lk.targetName());
		}
	}

	public static void removeMember(String groupName, ArrayList<Link> membersToRemove) {
		try {
			GroupAccessControlManager acm = new GroupAccessControlManager(null, groupStorage, userStorage, CCNHandle.open());
			GroupManager gm = acm.groupManager();
			Thread.sleep(TIMEOUT);
			Group g = gm.getGroup(groupName, SystemConfiguration.getDefaultTimeout());
			g.modify(null, membersToRemove);
		}
		catch (AccessDeniedException aed) {
			System.out.println("You do not have the permission to edit the membership of Group " + groupName);
			System.exit(1);
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.out.println("Removed from group " + groupName + " the following members: ");
		for (Link lk: membersToRemove) {
			System.out.println(lk.targetName());
		}
	}

}
